fastcrypto-zkp/src/groth16/generic_api.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::Deserialize;
+
+use fastcrypto::error::{FastCryptoError, FastCryptoResult};
+use fastcrypto::groups::{GroupElement, Pairing};
+use fastcrypto::serde_helpers::{deserialize_vector, serialize_vector, ToFromByteArray};
+
+use crate::groth16::{PreparedVerifyingKey, VerifyingKey};
+
+/// Deserialize bytes as an Arkworks representation of a verifying key, and return a vector of the
+/// four components of a prepared verified key (see more at [`PreparedVerifyingKey`]).
+pub(crate) fn prepare_pvk_bytes<
+    G1,
+    const G1_SIZE: usize,
+    const G2_SIZE: usize,
+    const GT_SIZE: usize,
+    const FR_SIZE: usize,
+>(
+    vk_bytes: &[u8],
+) -> FastCryptoResult<Vec<Vec<u8>>>
+where
+    G1: Pairing + for<'a> Deserialize<'a> + ToFromByteArray<G1_SIZE>,
+    <G1 as Pairing>::Other: for<'a> Deserialize<'a> + ToFromByteArray<G2_SIZE>,
+    <G1 as Pairing>::Output: GroupElement + ToFromByteArray<GT_SIZE>,
+{
+    let vk = VerifyingKey::<G1>::from_arkworks_format::<G1_SIZE, G2_SIZE>(vk_bytes)?;
+    Ok(PreparedVerifyingKey::from(&vk).serialize_into_parts())
+}
+
+/// Verify Groth16 proof using the serialized form of the four components in a prepared verifying key
+/// (see more at [`PreparedVerifyingKey`]), serialized proof public input, which should
+/// be concatenated serialized field elements of the scalar field of [`crate::conversions::SCALAR_SIZE`]
+/// bytes each, and serialized proof points.
+pub(crate) fn verify_groth16_in_bytes<
+    G1,
+    const G1_SIZE: usize,
+    const G2_SIZE: usize,
+    const GT_SIZE: usize,
+    const FR_SIZE: usize,
+>(
+    vk_gamma_abc_g1_bytes: &[u8],
+    alpha_g1_beta_g2_bytes: &[u8],
+    gamma_g2_neg_pc_bytes: &[u8],
+    delta_g2_neg_pc_bytes: &[u8],
+    proof_public_inputs_as_bytes: &[u8],
+    proof_points_as_bytes: &[u8],
+) -> Result<bool, FastCryptoError>
```

Comment:
we usually don't return bool - why not return InvalidProof in case the proof is invalid?